sql-plugin/src/main/scala/org/apache/spark/sql/rapids/GpuFileFormatDataWriter.scala

```diff
@@ -438,192 +423,164 @@ class GpuDynamicPartitionDataSingleWriter(
-            withResource(concat){concat =>  
-              writeBatch(concat)
-            }
-          }
-        } else {
-          if (!needSplitBatch(
-            maxRecordsPerFile,
-            currentWriterStatus.recordsInFile,
-            batch.numRows))
-          {
-            statsTrackers.foreach(_.newBatch(batch))
-            currentWriterStatus.recordsInFile += batch.numRows()
-            currentWriterStatus.outputWriter.write(batch, statsTrackers)
-          } else {
-            withResource(batch){batch =>  
-              writeBatch(batch)
+          closeOnExcept(GpuColumnVector.from(table, outDataTypes)) { batch =>
+            if (savedStatus.isDefined && savedStatus.get.tableCaches.nonEmpty) {
+              // convert caches seq to tables and close caches seq
+              val subTables = convertSpillBatchesToTables(savedStatus.get.tableCaches)
+              // clear the caches
+              savedStatus.get.tableCaches.clear()
+              // concat the caches and this `batch`
+              val concat = withResource(subTables) { _ =>
+                withResource(batch) { _ =>
+                  // append `batch` to sub tables
+                  subTables += GpuColumnVector.from(batch)
+                }
+                withResource(Table.concatenate(subTables: _*)) { concat =>
+                  GpuColumnVector.from(concat, outDataTypes)
+                }
+              }
+              // write concat table
+              if (!needSplitBatch(
+                maxRecordsPerFile,
+                currentWriterStatus.recordsInFile,
+                concat.numRows))
+              {
+                statsTrackers.foreach(_.newBatch(concat))
+                currentWriterStatus.recordsInFile += concat.numRows()
+                currentWriterStatus.outputWriter.write(concat, statsTrackers)
+              } else {
+                withResource(concat){ _ =>
+                  writeBatch(concat)
+                }
+              }
+            } else {
+              if (!needSplitBatch(
+                maxRecordsPerFile,
+                currentWriterStatus.recordsInFile,
+                batch.numRows))
+              {
+                statsTrackers.foreach(_.newBatch(batch))
+                currentWriterStatus.recordsInFile += batch.numRows()
+                currentWriterStatus.outputWriter.write(batch, statsTrackers)
+              } else {
+                withResource(batch){ _ =>
+                  writeBatch(batch)
+                }
+              }
             }
           }
-        }
-      })
-    } finally {
-      if (needToCloseBatch) {
-        cb.close()
-      }
-
-      if (partitionColumns != null) {
-        partitionColumns.close()
-      }
-
-      if (distinctKeys != null) {
-        distinctKeys.close()
-      }
-
-      if (outputColumns != null) {
-        outputColumns.close()
-      }
-
-      splits.safeClose()
-
-      if (cbKeys != null) {
-        cbKeys.close()
+        }}
```

Comment:
Only one `}` is needed if removing the tailing `{` at line 454

Reply:
Removed